mars/dataframe/expressions/arithmetic/core.py

```diff
@@ -0,0 +1,358 @@
+
+    def calc_shape(self, *inputs_shape):
+        return self.outputs[0].shape
+
+
+class DataFrameBinOpMixin(DataFrameOperandMixin):
+    __slots__ = ()
+
+    @classmethod
+    def _check_overlap(cls, chunk_index_min_max):
+        for j in range(len(chunk_index_min_max) - 1):
+            # overlap only if the prev max is close and curr min is close
+            # and they are identical
+            prev_max, prev_max_close = chunk_index_min_max[j][2:]
+            curr_min, curr_min_close = chunk_index_min_max[j + 1][:2]
+            if prev_max_close and curr_min_close and prev_max == curr_min:
+                return True
+        return False
+
+    @classmethod
+    def _get_chunk_index_min_max(cls, df, index_type, axis):
+        index = getattr(df, index_type)
+        if not index.is_monotonic_increasing_or_decreasing:
+            return
+
+        chunk_index_min_max = []
+        for i in range(df.chunk_shape[axis]):
+            chunk_idx = [0, 0]
+            chunk_idx[axis] = i
+            chunk = df.cix[tuple(chunk_idx)]
+            chunk_index = getattr(chunk, index_type)
+            min_val = chunk_index.min_val
+            min_val_close = chunk_index.min_val_close
+            max_val = chunk_index.max_val
+            max_val_close = chunk_index.max_val_close
+            if min_val is None or max_val is None:
+                return
+            chunk_index_min_max.append((min_val, min_val_close, max_val, max_val_close))
+
+        if index.is_monotonic_decreasing:
+            return list(reversed(chunk_index_min_max)), False
+
+        if cls._check_overlap(chunk_index_min_max):
+            return
+        return chunk_index_min_max, True
+
+    @classmethod
+    def _get_dtypes(cls, dtypes, column_min_max):
+        l_filter = operator.ge if column_min_max[1] else operator.gt
+        l = l_filter(dtypes.index, column_min_max[0])
+        r_filter = operator.le if column_min_max[3] else operator.lt
+        r = r_filter(dtypes.index, column_min_max[2])
+        f = l & r
+        return dtypes[f]
+
+    @classmethod
+    def _gen_out_chunks_without_shuffle(cls, op, splits, out_shape, left, right):
+        out_chunks = []
+        for out_idx in itertools.product(*(range(s) for s in out_shape)):
+            if splits[0] is not None:
+                # index does not need shuffle
+                left_row_idx, left_row_inner_idx = splits[0][0][1][out_idx[0]]
+                right_row_idx, right_row_inner_idx = splits[0][1][1][out_idx[0]]
+            else:
+                left_row_idx = left_row_inner_idx = \
+                    right_row_idx = right_row_inner_idx = None
+            if splits[1] is not None:
+                # column does not need shuffle
+                left_column_idx, left_column_inner_idx = splits[1][0][1][out_idx[1]]
+                right_column_idx, right_column_inner_idx = splits[1][1][1][out_idx[1]]
+            else:
+                left_column_idx = left_column_inner_idx = \
```

Comment:
Will these ever happen when we have already detected that no shuffle is needed? I also notice that this line is not covered in coverage report.

Reply:
ditto